### PR TITLE
aui2s: add rtos i2s audio driver module

### PR DESCRIPTION
**Pull request by cspiel1**
An I2S audio driver for freeRTOS.

Tested with:
MAX98357A - i2S CLASS D stereo amplifier
Sparkfun I2S Audio Breakout - MAX98357A SF14809

### CODE REVIEW

**Comment by cspiel1** [comment]
Builds only for freeRTOS.

We didn't use the baresip makefiles. Instead built it by writing component.mk files for esp-idf.
- https://github.com/mrxa13/esp32-baresip/tree/master/esp32phone/components/re_mk
- https://github.com/mrxa13/esp32-baresip/tree/master/esp32phone/components/rem_mk
- https://github.com/mrxa13/esp32-baresip/tree/master/esp32phone/components/baresip_mk

File: modules/aui2s/aui2s.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+			      "aui2s", aui2s_src_alloc);
+	err |= auplay_register(&auplay, baresip_auplayl(),
+			       "aui2s", aui2s_play_alloc);
+
+
+	return err;
+}
+
+
+static enum I2SOnMask _i2s_on = I2O_NONE;
```
Comment (alfredh) [inline modules/aui2s/aui2s.c]: move static variable to the top of the file...

File: modules/aui2s/aui2s.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+
+int aui2s_start(uint32_t srate, enum I2SOnMask playrec)
```
Comment (alfredh) [inline modules/aui2s/aui2s.c]: what about channels ? does it support multiple channels ?

File: modules/aui2s/aui2s.c [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		i2s_set_pin(I2S_PORT, &pins);
+
+		i2s_zero_dma_buffer(I2S_PORT);
+	}
+
+	return 0;
+}
+
+
+int aui2s_stop(enum I2SOnMask playrec)
```
Comment (alfredh) [inline modules/aui2s/aui2s.c]: return void instead, since it cannot fail

File: modules/aui2s/aui2s.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+	info("%s _i2s_on=%d", __func__, _i2s_on);
+	if (_i2s_on == I2O_NONE)
+		i2s_driver_uninstall(I2S_PORT);
+
+	return 0;
+}
+
+
+static int aui2s_close(void)
```
Comment (alfredh) [inline modules/aui2s/aui2s.c]: return void instead

File: modules/aui2s/aui2s_play.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	struct auplay_st *st = arg;
+	int err;
+
+	err = aui2s_start(st->prm.srate, I2O_PLAY);
+	if (err) {
+		warning("aui2s: could not start auplay\n");
+		return NULL;
+	}
+
+	i2s_set_clk(I2S_PORT, st->prm.srate, 32, 1);
```
Comment (alfredh) [inline modules/aui2s/aui2s_play.c]: the last parameter (1) is number of channels. perhaps make a constant for that,
and also check number of channels in the alloc functions.

**Comment by alfredh** [comment]
I suggest renaming the module to just `i2s`, since the I2S bus
is mainly for audio (Inter-IC Sound)

File: modules/aui2s/aui2s.c [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+			.mode = I2S_MODE_MASTER | I2S_MODE_RX | I2S_MODE_TX,
+			.sample_rate =  srate,
+			.bits_per_sample = 32,
+			.communication_format = I2S_COMM_FORMAT_I2S |
+				I2S_COMM_FORMAT_I2S_MSB,
+			.channel_format = I2S_CHANNEL_FMT_ONLY_RIGHT,
+			.intr_alloc_flags = ESP_INTR_FLAG_LEVEL1,
+			.dma_buf_count = 2,
+			.dma_buf_len = DMA_SIZE,
+			.use_apll = 0                       /* APPL DISABLE */
```
Comment (alfredh) [inline modules/aui2s/aui2s.c]: typo in comment, should be "APLL" ?

File: modules/aui2s/aui2s_play.c [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+#include "aui2s.h"
+
+
+struct auplay_st {
+	const struct auplay *ap;  /* pointer to base-class (inheritance) */
+	pthread_t thread;
+	bool run;
+	void *sampv;
+	size_t sampc;
+	size_t bytes;
```
Comment (alfredh) [inline modules/aui2s/aui2s_play.c]: bytes is unused ?

File: modules/aui2s/aui2s_play.c [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+		st->run = false;
+		(void)pthread_join(st->thread, NULL);
+	}
+
+	mem_deref(st->sampv);
+	mem_deref(st->pcm);
+}
+
+
+static void convert_sampv(struct auplay_st *st, size_t i, size_t n)
```
Comment (alfredh) [inline modules/aui2s/aui2s_play.c]: perhaps add a note about what this function is doing ..

File: modules/aui2s/aui2s_src.c [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+		return NULL;
+	}
+
+	while (st->run) {
+		size_t i;
+
+		for (i = 0; i + DMA_SIZE / 4 <= st->sampc;) {
+			size_t n = 0;
+			i2s_read(I2S_PORT, st->pcm, DMA_SIZE, &n,
+					portMAX_DELAY);
```
Comment (alfredh) [inline modules/aui2s/aui2s_src.c]: check return value of `i2s_read ` ?

File: modules/aui2s/aui2s.c [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	info("%s start with _i2s_on=%d", __func__, _i2s_on);
+	if (start) {
+		i2s_config_t i2s_config = {
+			.mode = I2S_MODE_MASTER | I2S_MODE_RX | I2S_MODE_TX,
+			.sample_rate =  srate,
+			.bits_per_sample = 32,
+			.communication_format = I2S_COMM_FORMAT_I2S |
+				I2S_COMM_FORMAT_I2S_MSB,
```
Comment (alfredh) [inline modules/aui2s/aui2s.c]: is it possible to use a communication format that is compatible
with the input from baresip core, i.e. signed 16-bit PCM ?

in this case, we can get rid of the sample format conversion...

File: modules/aui2s/aui2s_src.c [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+	}
+}
+
+
+static void *read_thread(void *arg)
+{
+	struct ausrc_st *st = arg;
+	int err;
+
+	err = aui2s_start(st->prm.srate, I2O_RECO);
```
Comment (alfredh) [inline modules/aui2s/aui2s_src.c]: is the I2S started from the thread on purpose ?

or can it be started from the alloc function?

(if aui2s_start fails, the thread will just exit but the program will continue
as if everything is working).

**Comment by cspiel1** [comment]
On Fri 08.11.19  23:45, Alfred E. Heggestad wrote:
> alfredh commented on this pull request.
> 
> 
> 
> > +	int err;
> +
> +	err  = ausrc_register(&ausrc, baresip_ausrcl(),
> +			      "aui2s", aui2s_src_alloc);
> +	err |= auplay_register(&auplay, baresip_auplayl(),
> +			       "aui2s", aui2s_play_alloc);
> +
> +
> +	return err;
> +}
> +
> +
> +static enum I2SOnMask _i2s_on = I2O_NONE;
> +
> +
> +int aui2s_start(uint32_t srate, enum I2SOnMask playrec)
> 
> what about channels ? does it support multiple channels ?

No, currently not. Our goal was a mono device. I tried several settings for
the channel_format and finally it worked with I2S_CHANNEL_FMT_ONLY_RIGHT.

Stereo should be also possible. E.g. using two MEMs mics one with channel
select pin on low and the other on high and I2S_CHANNEL_FMT_RIGHT_LEFT.

I could add this setting in the start function, but currently have no setup for
 testing.

r,c

**Comment by cspiel1** [comment]
Oh sorry. There will follow more corrections and I will make an interactive rebase, sure.
r,c

**Comment by cspiel1** [comment]
On Fri 08.11.19  23:58, Alfred E. Heggestad wrote:
> I suggest renaming the module to just `i2s`, since the I2S bus
> is mainly for audio (Inter-IC Sound)

Ok. I will rename it. But it is only for freeRTOS.

r,c

**Comment by cspiel1** [comment]
On Sat 09.11.19  00:12, Alfred E. Heggestad wrote:
> alfredh commented on this pull request.
> 
> 
> 
> > +	int err;
> +
> +	err = aui2s_start(st->prm.srate, I2O_RECO);
> +
> +	if (err) {
> +		warning("aui2s: could not start ausrc\n");
> +		return NULL;
> +	}
> +
> +	while (st->run) {
> +		size_t i;
> +
> +		for (i = 0; i + DMA_SIZE / 4 <= st->sampc;) {
> +			size_t n = 0;
> +			i2s_read(I2S_PORT, st->pcm, DMA_SIZE, &n,
> +					portMAX_DELAY);
> 
> check return value of `i2s_read ` ?
> 

In the current implementation from esp-idf always ESP_OK is returned. But to be
save, I will add this check.

r,c

**Comment by cspiel1** [comment]
On Sat 09.11.19  04:51, Alfred E. Heggestad wrote:
> alfredh commented on this pull request.
> 
> 
> 
> > +	bool start = _i2s_on == I2O_NONE;
> +	_i2s_on = _i2s_on | playrec;
> +
> +	if (srate * 4 % DMA_SIZE) {
> +		warning("aui2s: srate*4 % DMA_SIZE != 0\n");
> +		return EINVAL;
> +	}
> +
> +	info("%s start with _i2s_on=%d", __func__, _i2s_on);
> +	if (start) {
> +		i2s_config_t i2s_config = {
> +			.mode = I2S_MODE_MASTER | I2S_MODE_RX | I2S_MODE_TX,
> +			.sample_rate =  srate,
> +			.bits_per_sample = 32,
> +			.communication_format = I2S_COMM_FORMAT_I2S |
> +				I2S_COMM_FORMAT_I2S_MSB,
> 
> is it possible to use a communication format that is compatible
> with the input from baresip core, i.e. signed 16-bit PCM ?
> 
> in this case, we can get rid of the sample format conversion...
> 

I tried to find this. But it did'nt work for the used hardware. I think the
SPH0645LM4H MEMs mics need 32 bit samples. And both, the speaker and the mic
are on the same bus.

Note: The shifting is in order to get a reasonable volume. I will add a
comment.

r,c

**Comment by cspiel1** [comment]
On Sat 09.11.19  04:55, Alfred E. Heggestad wrote:
> alfredh commented on this pull request.
> 
> 
> 
> > +		uint32_t v = st->pcm[j];
> +		uint16_t *o = sampv + i + j;
> +		*o = v >> 15;
> +		/* if negative fill with ff */
> +		if (v & 0x80000000)
> +			*o |= 0xfffe0000;
> +	}
> +}
> +
> +
> +static void *read_thread(void *arg)
> +{
> +	struct ausrc_st *st = arg;
> +	int err;
> +
> +	err = aui2s_start(st->prm.srate, I2O_RECO);
> 
> is the I2S started from the thread on purpose ?
> 
> or can it be started from the alloc function?
> 
> (if aui2s_start fails, the thread will just exit but the program will continue
> as if everything is working).
> 

You are right. It is better to try first i2s_start, then create the thread.

r,c

**Comment by cspiel1** [comment]
Ok. Did rework due to Alfred's comments. But we have to test it again on our esp32 phone.

**Comment by cspiel1** [comment]
Ok, now we tested the module in the project https://github.com/mrxa13/esp32-baresip. Works like before the rework.

**Comment by alfredh** [comment]
great, looks really good now.

is it okay to add you as the maintainer of the new i2s module?

https://github.com/alfredh/baresip/wiki/Modules

**Comment by cspiel1** [comment]
Yes, of course. Thanks!

r,chris

**Comment by cspiel1** [comment]
On Mon 11.11.19  05:38, Alfred E. Heggestad wrote:
> great, looks really good now.
> 
> is it okay to add you as the maintainer of the new i2s module?
> 
> https://github.com/alfredh/baresip/wiki/Modules

Hi Alfred,

thank you for the merge! But sorry, I made a mistake in the commit message.


    Tested with:
    MAX98357A - i2S CLASS D stereo amplifier
    Sparkfun I2S Audio Breakout - MAX98357A SF14809


This is the amplifier twice. No hint about the MEMs Mics. Maybe we should add
a short description in the head of i2s.h.


Tested with:
Sparkfun I2S Audio Breakout - MAX98357A SF14809 - CLASS D stereo amplifier
SPH0645LM4H - Mems i2s stereo mic


r,c

**Comment by alfredh** [comment]
it is probably best to add a README.md file in modules/i2s
or add a doxygen header like this:

https://github.com/alfredh/baresip/blob/master/modules/ice/ice.c#L14

**Pull request merged**